src/mlpack/methods/ann/layer/flexible_relu_impl.hpp

```diff
@@ -0,0 +1,87 @@
+/**
+ * @file flexible_relu_impl.hpp
+ * @author Aarush Gupta
+ * @author Manthan-R-Sheth
+ *
+ * Implementation of FlexibleReLU layer as described by
+ * Suo Qiu, Xiangmin Xu and Bolun Cai in
+ * "FReLU: Flexible Rectified Linear Units for Improving Convolutional 
+ *  Neural Networks", 2018
+ *  
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_ANN_LAYER_FLEXIBLERELU_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_FLEXIBLERELU_IMPL_HPP
+
+#include "flexible_relu.hpp"
+#include<algorithm>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InputDataType, typename OutputDataType>
+FlexibleReLU<InputDataType, OutputDataType>::FlexibleReLU(
+    const double userAlpha) : userAlpha(userAlpha)
+{
+  alpha.set_size(1, 1);
+  alpha(0) = userAlpha;
+}
+
+template<typename InputDataType, typename OutputDataType>
+void FlexibleReLU<InputDataType, OutputDataType>::Reset()
+{
+  //! Set value of alpha to the one given by user.
+  alpha(0) = userAlpha;
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename InputType, typename OutputType>
+void FlexibleReLU<InputDataType, OutputDataType>::Forward(
+    const InputType&& input, OutputType&& output)
+{
+  output = arma::clamp(input, 0.0, DBL_MAX) + alpha(0);
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename DataType>
+void FlexibleReLU<InputDataType, OutputDataType>::Backward(
+    const DataType&& input, DataType&& gy, DataType&& g)
+{
+  DataType derivative;
+  //! Compute the first derivative of FlexibleReLU function.
+  derivative = arma::sign(input);
+  derivative.elem(arma::find(derivative < 0.0)) += 1;
```

Comment:
We could do something like: ``derivative.elem(find(input > 0) ).ones();`` here, I think it's easier to get. Let me know what you think.

Reply:
i think i will have to initialize ```derivative``` with ```zeros()``` or ```input``` before using ```derivative.elem(find(input > 0) ).ones();```.
Tell me if i missed anything.

Reply:
you are right

Reply:
It would be possible to also do `derivative = arma::clamp(arma::sign(input), 0.0, 1.0);` but I don't think this will currently give any acceleration. It may be slightly easier to read though.